library/alloc/src/collections/vec_deque/iter.rs

```diff
@@ -13,76 +10,72 @@ use super::{count, wrap_index, RingSlices};
 /// [`iter`]: super::VecDeque::iter
 #[stable(feature = "rust1", since = "1.0.0")]
 pub struct Iter<'a, T: 'a> {
-    ring: &'a [MaybeUninit<T>],
-    tail: usize,
-    head: usize,
+    i1: slice::Iter<'a, T>,
+    i2: slice::Iter<'a, T>,
 }
 
 impl<'a, T> Iter<'a, T> {
-    pub(super) fn new(ring: &'a [MaybeUninit<T>], tail: usize, head: usize) -> Self {
-        Iter { ring, tail, head }
+    pub(super) fn new(i1: slice::Iter<'a, T>, i2: slice::Iter<'a, T>) -> Self {
+        Self { i1, i2 }
     }
 }
 
 #[stable(feature = "collection_debug", since = "1.17.0")]
 impl<T: fmt::Debug> fmt::Debug for Iter<'_, T> {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        let (front, back) = RingSlices::ring_slices(self.ring, self.head, self.tail);
-        // Safety:
-        // - `self.head` and `self.tail` in a ring buffer are always valid indices.
-        // - `RingSlices::ring_slices` guarantees that the slices split according to `self.head` and `self.tail` are initialized.
-        unsafe {
-            f.debug_tuple("Iter")
-                .field(&MaybeUninit::slice_assume_init_ref(front))
-                .field(&MaybeUninit::slice_assume_init_ref(back))
-                .finish()
-        }
+        f.debug_tuple("Iter").field(&self.i1.as_slice()).field(&self.i2.as_slice()).finish()
     }
 }
 
 // FIXME(#26925) Remove in favor of `#[derive(Clone)]`
 #[stable(feature = "rust1", since = "1.0.0")]
 impl<T> Clone for Iter<'_, T> {
     fn clone(&self) -> Self {
-        Iter { ring: self.ring, tail: self.tail, head: self.head }
+        Iter { i1: self.i1.clone(), i2: self.i2.clone() }
     }
 }
 
 #[stable(feature = "rust1", since = "1.0.0")]
 impl<'a, T> Iterator for Iter<'a, T> {
     type Item = &'a T;
 
+    fn advance_by(&mut self, n: usize) -> Result<(), usize> {
+        let m = match self.i1.advance_by(n) {
+            Ok(_) => return Ok(()),
+            Err(m) => m,
+        };
+        mem::swap(&mut self.i1, &mut self.i2);
+        self.i1.advance_by(n - m).map_err(|o| o + m)
+    }
+
     #[inline]
     fn next(&mut self) -> Option<&'a T> {
-        if self.tail == self.head {
-            return None;
+        match self.i1.next() {
+            Some(val) => Some(val),
+            None => {
+                // most of the time, the iterator will either always
+                // call next(), or always call next_back(). By swapping
+                // the iterators once the first one is empty, we ensure
+                // that the first branch is taken as often as possible,
+                // without sacrificing correctness, as i1 is empty anyways
+                mem::swap(&mut self.i1, &mut self.i2);
```

Comment:
I guess this is the same "are you planning to change these iterators?" kind of question, but it's not obvious to me that this is necessarily the best way to handle these.  Do you have anything to show that LLVM handles external iteration better with this swap in place, vs the simple impl of just `self.i1.next().or_else(|| self.i2.next())`?  Having a swap seems like it'd make loop canonicalization/splitting hard.

But maybe current LLVM can't actually split nor unroll such looks today, and then this is helpful?

Reply:
I also wonder about how it would impact `while let (Some(a), Some(b)) = (it.next(), it.next_back()) { … }`-style loops.  If only one of the inner slice iterators has values (like is common in a `collect`ed `VecDeque`) would that mean it'd swap twice every iteration of that loop?

Reply:
Admittedly, I haven't thought about it too much, other than "fewer branches probably better". I can try to set up a benchmark for it to see what kind of impact the swapping actually has. In my mind, the `(it.next(), it.next_back())` was sort of a pathological case, as I can't think of many cases where one would need it, but I'm open to changing it.

Reply:
Well I did just some very quick tests (replacing the swap with just `self.i2.next()` and running the `bench_iter_1000` benchmark) and without the swap it's consistently roughly 2x slower in that benchmark. I don't really get why this is happening though, as the deque in that benchmark should just be one contiguous slice anyways. It tried it enough times to rule out bad scheduling though, so for now I'll just leave the swaps in.

Reply:
`(it.next(), it.next_back())` on `IterMut` is what I'd do if I wanted a `VecDeque::reverse`, for example.

I don't think this needs to block the PR, though.  It's probably fine as-is.